// res-pareto-filesystem/dist/imp/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWriteStream = exports.mkdir = exports.getFile = exports.unlink = exports.writeFile = exports.readDirectory = exports.readFile = void 0;
var readFile_1 = require("./functions/readFile");
Object.defineProperty(exports, "readFile", { enumerable: true, get: function () { return readFile_1.readFile; } });
var readDirectory_1 = require("./functions/readDirectory");
Object.defineProperty(exports, "readDirectory", { enumerable: true, get: function () { return readDirectory_1.readDirectory; } });
var writeFile_1 = require("./functions/writeFile");
Object.defineProperty(exports, "writeFile", { enumerable: true, get: function () { return writeFile_1.writeFile; } });
var unlink_1 = require("./functions/unlink");
Object.defineProperty(exports, "unlink", { enumerable: true, get: function () { return unlink_1.unlink; } });
var getFile_1 = require("./functions/getFile");
Object.defineProperty(exports, "getFile", { enumerable: true, get: function () { return getFile_1.getFile; } });
var mkDir_1 = require("./functions/mkDir");
Object.defineProperty(exports, "mkdir", { enumerable: true, get: function () { return mkDir_1.mkdir; } });
var createWriteStream_1 = require("./functions/createWriteStream");
Object.defineProperty(exports, "createWriteStream", { enumerable: true, get: function () { return createWriteStream_1.createWriteStream; } });
